### PR TITLE
Raven db 9479

### PR DESCRIPTION
**Pull request by talweiss1982**
This PR is for reviewing the changes while i continue to debug NewtonSoft Behavior under ravendb

### CODE REVIEW

File: src/Sparrow/Json/BlittableJsonTextWriter.cs [hunk 1]
```diff
@@ -247,10 +247,9 @@ public void WriteString(LazyStringValue str, bool skipEscaping = false)
 
             var size = str.Size;
 
-            if (size == 1 && str.IsControlCodeCharacter(out var b))
+            if (size == 1 && char.IsControl((char)str[0]))
```
Comment (ayende) [inline src/Sparrow/Json/BlittableJsonTextWriter.cs:250]: This code needs to be removed

File: src/Sparrow/Json/Parsing/JsonParserState.cs [hunk 2]
```diff
@@ -104,10 +116,33 @@ public static void FindEscapePositionsIn(FastList<int> buffer, byte* str, int le
...
+                if (value < 32 || value >= 127 && value <= 159)
+                {
+                    // move rest of buffer 
+                    // write \u0000
+                    // update size
+                    var from = str + i + 1;
+                    var to = str + i + 1 + SizeOfUtf8AsStringInBytesWithoutEscaping;
+                    var sizeToCopy = len - i;
+                    //here we only shifting by 5 bytes since we are going to override the byte at the current position.
+                    Memory.Copy(to, from, sizeToCopy ); //TODO: we must verify that we have enough space in our buffer otherwise we must request a bigger one
```
Comment (ayende) [inline src/Sparrow/Json/Parsing/JsonParserState.cs:132]: We shouldn't request more, but should validate that we aren't over the max size

File: src/Sparrow/Json/Parsing/JsonParserState.cs [hunk 3]
```diff
@@ -131,5 +166,37 @@ public void Reset()
             Continuation = JsonParserTokenContinuation.None;
             EscapePositions.Clear();
         }
+
+        //This table was programaticly generated 
+        public static readonly byte[][] EscapeCharactersStringsAsBytes = new byte[160][]
```
Comment (ayende) [inline src/Sparrow/Json/Parsing/JsonParserState.cs:171]: Generate this at runtime, instead. This is not going to be readable long term.

File: src/Sparrow/Json/Parsing/JsonParserState.cs (hunk 2 above)
Comment (maximburyak) [reply src/Sparrow/Json/Parsing/JsonParserState.cs:132]: @talweiss1982 that todo remark is very important! I think that we may want to perform that modification before that function, moreover, we should first count that amount of "control charachters", according to that, to allocate or reallocate the right memory chunk and only then to perform that code

Comment (maximburyak) [reply src/Sparrow/Json/Parsing/JsonParserState.cs:132]: So after talking with @talweiss1982 , it seems that I've overlooked the modifications to FindEscapePositionsMaxSize, It wtill seems a bit awkward to modify data inside FindEscapePositionsIn, but as an intermidiate solution, it does the work

File: src/Sparrow/Json/Parsing/JsonParserState.cs [hunk 4]
```diff
@@ -104,10 +116,33 @@ public static void FindEscapePositionsIn(FastList<int> buffer, byte* str, int le
...
+                    // move rest of buffer 
+                    // write \u0000
+                    // update size
+                    var from = str + i + 1;
+                    var to = str + i + 1 + SizeOfUtf8AsStringInBytesWithoutEscaping;
+                    var sizeToCopy = len - i;
+                    //here we only shifting by 5 bytes since we are going to override the byte at the current position.
+                    Memory.Copy(to, from, sizeToCopy ); //TODO: we must verify that we have enough space in our buffer otherwise we must request a bigger one
+                    //TODO: find a way to generate the table as pinned
+                    fixed (byte* controlString = EscapeCharactersStringsAsBytes[value])
```
Comment (maximburyak) [inline src/Sparrow/Json/Parsing/JsonParserState.cs:134]: it's GCHandle.Allocate

**Pull request closed without merging**